File: packages/agent-sdk/src/submissions.ts
import type { CaseId } from "./cases";
import type { CaseTypeTemplateId, DomainPackId } from "./domains";
import type { PacketId, PacketSectionType } from "./packets";
import type { ReadinessStatus } from "./readiness";
import type { DownstreamSourceMode, SignOffStatus } from "./reviewed-handoff";
import type {
  CaseTargetPackSelection,
  TargetAutomationCompatibility,
} from "./target-packs";

export type SubmissionTargetId = string;
export type SubmissionDraftId = string;
export type SubmissionMappingId = string;
export type AutomationPlanId = string;

export type AutomationExecutionMode =
  | "playwright_mcp"
  | "computer_use_fallback"
  | "manual_only"
  | "blocked";

export type SubmissionTargetCategory =
  | "portal_submission"
  | "form_packet_export"
  | "internal_handoff_packet";

export type SubmissionDraftStatus =
  | "draft_created"
  | "mapping_incomplete"
  | "awaiting_operator_review"
  | "approved_for_future_execution"
  | "blocked"
  | "superseded_placeholder";

export type SubmissionSourceEntityType =
  | "case"
  | "case_metadata"
  | "packet_manifest"
  | "packet_section"
  | "document"
  | "extraction"
  | "readiness"
  | "reviewed_snapshot";

export type SubmissionMappingStatus =
  | "unresolved"
  | "candidate_available"
  | "mapped_preview"
  | "requires_human_input";

export type AutomationPlanStepType =
  | "open_target"
  | "navigate_section"
  | "populate_field_placeholder"
  | "attach_document_placeholder"
  | "review_before_submit"
  | "submit_blocked_placeholder";

export type AutomationPlanStatus =
  | "draft"
  | "partial"
  | "awaiting_operator_review"
  | "approved_for_future_execution"
  | "blocked";

export type AutomationPlanStepStatus =
  | "informational"
  | "future_automation_placeholder"
  | "requires_human_input"
  | "blocked";

export type ApprovalStatus =
  | "not_requested"
  | "awaiting_operator_review"
  | "approved_for_future_execution"
  | "rejected";

export type ResultIssueSeverity = "info" | "warning" | "error";

export interface SubmissionMappingSourceReference {
  source_entity_type: SubmissionSourceEntityType;
  source_entity_id: string;
  source_path: string;
  display_label: string;
}

export interface SubmissionMappingTargetField {
  field_name: string;
  target_section: string;
  display_label: string;
  field_type: string;
  required: boolean;
  candidate_source_paths: string[];
  notes: string[];
}

export interface SubmissionFieldValuePreview {
  value_present: boolean;
  text_value: string;
  raw_value: unknown;
  source_reference: SubmissionMappingSourceReference | null;
  notes: string[];
}

export interface SubmissionMappingFieldDefinition {
  mapping_id: SubmissionMappingId;
  target_field: SubmissionMappingTargetField;
  status: SubmissionMappingStatus;
  source_reference: SubmissionMappingSourceReference | null;
  value_preview: SubmissionFieldValuePreview | null;
  notes: string[];
}

export interface SubmissionTargetMetadata {
  target_id: SubmissionTargetId;
  category: SubmissionTargetCategory;
  display_name: string;
  description: string;
  notes: string[];
  supported_domain_pack_ids: DomainPackId[];
  supported_case_type_ids: CaseTypeTemplateId[];
  supports_field_mapping: boolean;
  supports_file_attachments: boolean;
  supports_dry_run_preview: boolean;
  supports_live_submission: boolean;
  default_backend_ids: string[];
  default_target_fields: SubmissionMappingTargetField[];
}

export interface SubmissionTargetListResponse {
  targets: SubmissionTargetMetadata[];
}

export interface SubmissionDraftSourceMetadata {
  packet_id: PacketId;
  source_mode: DownstreamSourceMode;
  source_reviewed_snapshot_id: string;
  source_snapshot_signoff_status: SignOffStatus;
  source_snapshot_signed_off_at: string;
  source_snapshot_signed_off_by: string;
  packet_generated_at: string;
  domain_pack_id: DomainPackId | null;
  case_type_id: CaseTypeTemplateId | null;
  readiness_status: ReadinessStatus | null;
  linked_document_count: number;
  extraction_count: number;
  candidate_source_count: number;
  source_sections: PacketSectionType[];
  target_pack_selection: CaseTargetPackSelection | null;
}

export interface ApprovalRequirementMetadata {
  requires_operator_approval: boolean;
  approval_status: ApprovalStatus;
  approved_by: string;
  approved_at: string;
  approval_note: string;
  scope: string;
}

export interface ExecutionGuardrailMetadata {
  requires_operator_approval: boolean;
  approval_status: ApprovalStatus;
  browser_write_actions_blocked: boolean;
  live_submission_blocked: boolean;
  allowed_backend_ids: string[];
  allowed_tool_ids: string[];
  blocked_actions: string[];
  notes: string[];
}

export interface AutomationFallbackRoutingHint {
  recommended_mode: AutomationExecutionMode;
  reason: string;
  supported_provider_ids: string[];
  notes: string[];
}

export interface NormalizedResultIssue {
  severity: ResultIssueSeverity;
  code: string;
  message: string;
  related_entity_type: string | null;
  related_entity_id: string | null;
}

export interface NormalizedOperationResult {
  success: boolean;
  message: string;
  issues: NormalizedResultIssue[];
}

export interface AutomationPlanStep {
  step_id: string;
  step_index: number;
  step_type: AutomationPlanStepType;
  status: AutomationPlanStepStatus;
  title: string;
  description: string;
  target_reference: string;
  tool_id: string | null;
  backend_id: string | null;
  execution_mode: AutomationExecutionMode;
  checkpoint_required: boolean;
  checkpoint_reason: string;
  fallback_hint: AutomationFallbackRoutingHint | null;
  mapping_id: SubmissionMappingId | null;
  related_document_id: string | null;
  notes: string[];
}

export interface DryRunResultSummary {
  plan_status: AutomationPlanStatus;
  total_steps: number;
  informational_steps: number;
  future_automation_steps: number;
  requires_human_input_steps: number;
  blocked_steps: number;
  missing_required_mapping_count: number;
  attachment_count: number;
  referenced_tool_count: number;
  referenced_backend_count: number;
  notes: string[];
}

export interface AutomationPlan {
  plan_id: AutomationPlanId;
  draft_id: SubmissionDraftId;
  target_id: SubmissionTargetId;
  target_pack_selection: CaseTargetPackSelection | null;
  target_pack_automation_compatibility: TargetAutomationCompatibility | null;
  source_mode: DownstreamSourceMode;
  source_reviewed_snapshot_id: string;
  status: AutomationPlanStatus;
  dry_run: boolean;
  generated_at: string;
  guardrails: ExecutionGuardrailMetadata;
  dry_run_summary: DryRunResultSummary;
  steps: AutomationPlanStep[];
}

export interface SubmissionDraftSummary {
  draft_id: SubmissionDraftId;
  case_id: CaseId;
  case_title: string;
  packet_id: PacketId;
  source_mode: DownstreamSourceMode;
  source_reviewed_snapshot_id: string;
  submission_target_id: SubmissionTargetId;
  submission_target_category: SubmissionTargetCategory;
  target_pack_selection: CaseTargetPackSelection | null;
  status: SubmissionDraftStatus;
  approval_status: ApprovalStatus;
  mapping_count: number;
  unresolved_mapping_count: number;
  created_at: string;
  updated_at: string;
  note: string;
}

export interface SubmissionDraftListResponse {
  drafts: SubmissionDraftSummary[];
}

export interface CreateSubmissionDraftRequest {
  packet_id: PacketId;
  submission_target_id: SubmissionTargetId;
  note?: string;
}

export interface SubmissionDraftCreateResponse {
  result: NormalizedOperationResult;
  draft: SubmissionDraftSummary;
  target: SubmissionTargetMetadata;
  source_metadata: SubmissionDraftSourceMetadata;
  mappings: SubmissionMappingFieldDefinition[];
  approval: ApprovalRequirementMetadata;
}

export interface SubmissionDraftDetailResponse {
  draft: SubmissionDraftSummary;
  target: SubmissionTargetMetadata;
  source_metadata: SubmissionDraftSourceMetadata;
  mappings: SubmissionMappingFieldDefinition[];
  approval: ApprovalRequirementMetadata;
  plan: AutomationPlan | null;
}

export interface GenerateAutomationPlanRequest {
  dry_run?: boolean;
}

export interface AutomationPlanGenerateResponse {
  result: NormalizedOperationResult;
  draft: SubmissionDraftSummary;
  plan: AutomationPlan;
}

export interface AutomationPlanResponse {
  plan: AutomationPlan;
}

export interface UpdateSubmissionApprovalRequest {
  approval_status: ApprovalStatus;
  approved_by?: string;
  approval_note?: string;
}

export interface SubmissionApprovalUpdateResponse {
  result: NormalizedOperationResult;
  draft: SubmissionDraftSummary;
  approval: ApprovalRequirementMetadata;
}

File: packages/agent-sdk/casegraph_agent_sdk/submissions.py
"""Shared submission draft and automation planning contracts.

This layer defines reviewable, approval-gated metadata between packet assembly
and any future live submission automation. It does not implement real portal
posting, payer-specific rules, regulatory filing logic, browser writes, or
 autonomous submission execution.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field

from casegraph_agent_sdk.cases import CaseId
from casegraph_agent_sdk.domains import CaseTypeTemplateId, DomainPackId
from casegraph_agent_sdk.packets import PacketId, PacketSectionType
from casegraph_agent_sdk.readiness import ReadinessStatus
from casegraph_agent_sdk.reviewed_handoff import DownstreamSourceMode, SignOffStatus
from casegraph_agent_sdk.target_packs import CaseTargetPackSelection, TargetAutomationCompatibility

SubmissionTargetId = str
SubmissionDraftId = str
SubmissionMappingId = str
AutomationPlanId = str

AutomationExecutionMode = Literal[
    "playwright_mcp",
    "computer_use_fallback",
    "manual_only",
    "blocked",
]

SubmissionTargetCategory = Literal[
    "portal_submission",
    "form_packet_export",
    "internal_handoff_packet",
]

SubmissionDraftStatus = Literal[
    "draft_created",
    "mapping_incomplete",
    "awaiting_operator_review",
    "approved_for_future_execution",
    "blocked",
    "superseded_placeholder",
]

SubmissionSourceEntityType = Literal[
    "case",
    "case_metadata",
    "packet_manifest",
    "packet_section",
    "document",
    "extraction",
    "readiness",
    "reviewed_snapshot",
]

SubmissionMappingStatus = Literal[
    "unresolved",
    "candidate_available",
    "mapped_preview",
    "requires_human_input",
]

AutomationPlanStepType = Literal[
    "open_target",
    "navigate_section",
    "populate_field_placeholder",
    "attach_document_placeholder",
    "review_before_submit",
    "submit_blocked_placeholder",
]

AutomationPlanStatus = Literal[
    "draft",
    "partial",
    "awaiting_operator_review",
    "approved_for_future_execution",
    "blocked",
]

AutomationPlanStepStatus = Literal[
    "informational",
    "future_automation_placeholder",
    "requires_human_input",
    "blocked",
]

ApprovalStatus = Literal[
    "not_requested",
    "awaiting_operator_review",
    "approved_for_future_execution",
    "rejected",
]

ResultIssueSeverity = Literal["info", "warning", "error"]


class SubmissionMappingSourceReference(BaseModel):
    source_entity_type: SubmissionSourceEntityType
    source_entity_id: str = ""
    source_path: str = ""
    display_label: str = ""


class SubmissionMappingTargetField(BaseModel):
    field_name: str
    target_section: str = ""
    display_label: str = ""
    field_type: str = "text"
    required: bool = False
    candidate_source_paths: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class SubmissionFieldValuePreview(BaseModel):
    value_present: bool = False
    text_value: str = ""
    raw_value: Any = None
    source_reference: SubmissionMappingSourceReference | None = None
    notes: list[str] = Field(default_factory=list)


class SubmissionMappingFieldDefinition(BaseModel):
    mapping_id: SubmissionMappingId
    target_field: SubmissionMappingTargetField
    status: SubmissionMappingStatus = "unresolved"
    source_reference: SubmissionMappingSourceReference | None = None
    value_preview: SubmissionFieldValuePreview | None = None
    notes: list[str] = Field(default_factory=list)


class SubmissionTargetMetadata(BaseModel):
    target_id: SubmissionTargetId
    category: SubmissionTargetCategory
    display_name: str
    description: str = ""
    notes: list[str] = Field(default_factory=list)
    supported_domain_pack_ids: list[DomainPackId] = Field(default_factory=list)
    supported_case_type_ids: list[CaseTypeTemplateId] = Field(default_factory=list)
    supports_field_mapping: bool = True
    supports_file_attachments: bool = True
    supports_dry_run_preview: bool = True
    supports_live_submission: bool = False
    default_backend_ids: list[str] = Field(default_factory=list)
    default_target_fields: list[SubmissionMappingTargetField] = Field(default_factory=list)


class SubmissionTargetListResponse(BaseModel):
    targets: list[SubmissionTargetMetadata] = Field(default_factory=list)


class SubmissionDraftSourceMetadata(BaseModel):
    packet_id: PacketId
    source_mode: DownstreamSourceMode = "live_case_state"
    source_reviewed_snapshot_id: str = ""
    source_snapshot_signoff_status: SignOffStatus = "not_signed_off"
    source_snapshot_signed_off_at: str = ""
    source_snapshot_signed_off_by: str = ""
    packet_generated_at: str = ""
    domain_pack_id: DomainPackId | None = None
    case_type_id: CaseTypeTemplateId | None = None
    readiness_status: ReadinessStatus | None = None
    linked_document_count: int = 0
    extraction_count: int = 0
    candidate_source_count: int = 0
    source_sections: list[PacketSectionType] = Field(default_factory=list)
    target_pack_selection: CaseTargetPackSelection | None = None


class ApprovalRequirementMetadata(BaseModel):
    requires_operator_approval: bool = True
    approval_status: ApprovalStatus = "not_requested"
    approved_by: str = ""
    approved_at: str = ""
    approval_note: str = ""
    scope: str = "future_execution"


class ExecutionGuardrailMetadata(BaseModel):
    requires_operator_approval: bool = True
    approval_status: ApprovalStatus = "not_requested"
    browser_write_actions_blocked: bool = True
    live_submission_blocked: bool = True
    allowed_backend_ids: list[str] = Field(default_factory=list)
    allowed_tool_ids: list[str] = Field(default_factory=list)
    blocked_actions: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class AutomationFallbackRoutingHint(BaseModel):
    recommended_mode: AutomationExecutionMode = "blocked"
    reason: str = ""
    supported_provider_ids: list[str] = Field(default_factory=list)
    notes: list[str] = Field(default_factory=list)


class NormalizedResultIssue(BaseModel):
    severity: ResultIssueSeverity = "info"
    code: str
    message: str
    related_entity_type: str | None = None
    related_entity_id: str | None = None


class NormalizedOperationResult(BaseModel):
    success: bool = True
    message: str = ""
    issues: list[NormalizedResultIssue] = Field(default_factory=list)


class AutomationPlanStep(BaseModel):
    step_id: str
    step_index: int = 0
    step_type: AutomationPlanStepType
    status: AutomationPlanStepStatus = "informational"
    title: str
    description: str = ""
    target_reference: str = ""
    tool_id: str | None = None
    backend_id: str | None = None
    execution_mode: AutomationExecutionMode = "blocked"
    checkpoint_required: bool = False
    checkpoint_reason: str = ""
    fallback_hint: AutomationFallbackRoutingHint | None = None
    mapping_id: SubmissionMappingId | None = None
    related_document_id: str | None = None
    notes: list[str] = Field(default_factory=list)


class DryRunResultSummary(BaseModel):
    plan_status: AutomationPlanStatus = "draft"
    total_steps: int = 0
    informational_steps: int = 0
    future_automation_steps: int = 0
    requires_human_input_steps: int = 0
    blocked_steps: int = 0
    missing_required_mapping_count: int = 0
    attachment_count: int = 0
    referenced_tool_count: int = 0
    referenced_backend_count: int = 0
    notes: list[str] = Field(default_factory=list)


class AutomationPlan(BaseModel):
    plan_id: AutomationPlanId
    draft_id: SubmissionDraftId
    target_id: SubmissionTargetId
    target_pack_selection: CaseTargetPackSelection | None = None
    target_pack_automation_compatibility: TargetAutomationCompatibility | None = None
    source_mode: DownstreamSourceMode = "live_case_state"
    source_reviewed_snapshot_id: str = ""
    status: AutomationPlanStatus = "draft"
    dry_run: bool = True
    generated_at: str = ""
    guardrails: ExecutionGuardrailMetadata
    dry_run_summary: DryRunResultSummary
    steps: list[AutomationPlanStep] = Field(default_factory=list)


class SubmissionDraftSummary(BaseModel):
    draft_id: SubmissionDraftId
    case_id: CaseId
    case_title: str = ""
    packet_id: PacketId
    source_mode: DownstreamSourceMode = "live_case_state"
    source_reviewed_snapshot_id: str = ""
    submission_target_id: SubmissionTargetId
    submission_target_category: SubmissionTargetCategory
    target_pack_selection: CaseTargetPackSelection | None = None
    status: SubmissionDraftStatus = "draft_created"
    approval_status: ApprovalStatus = "not_requested"
    mapping_count: int = 0
    unresolved_mapping_count: int = 0
    created_at: str = ""
    updated_at: str = ""
    note: str = ""


class SubmissionDraftListResponse(BaseModel):
    drafts: list[SubmissionDraftSummary] = Field(default_factory=list)


class CreateSubmissionDraftRequest(BaseModel):
    packet_id: PacketId
    submission_target_id: SubmissionTargetId
    note: str = ""


class SubmissionDraftCreateResponse(BaseModel):
    result: NormalizedOperationResult
    draft: SubmissionDraftSummary
    target: SubmissionTargetMetadata
    source_metadata: SubmissionDraftSourceMetadata
    mappings: list[SubmissionMappingFieldDefinition] = Field(default_factory=list)
    approval: ApprovalRequirementMetadata


class SubmissionDraftDetailResponse(BaseModel):
    draft: SubmissionDraftSummary
    target: SubmissionTargetMetadata
    source_metadata: SubmissionDraftSourceMetadata
    mappings: list[SubmissionMappingFieldDefinition] = Field(default_factory=list)
    approval: ApprovalRequirementMetadata
    plan: AutomationPlan | None = None


class GenerateAutomationPlanRequest(BaseModel):
    dry_run: bool = True


class AutomationPlanGenerateResponse(BaseModel):
    result: NormalizedOperationResult
    draft: SubmissionDraftSummary
    plan: AutomationPlan


class AutomationPlanResponse(BaseModel):
    plan: AutomationPlan


class UpdateSubmissionApprovalRequest(BaseModel):
    approval_status: ApprovalStatus
    approved_by: str = ""
    approval_note: str = ""


class SubmissionApprovalUpdateResponse(BaseModel):
    result: NormalizedOperationResult
    draft: SubmissionDraftSummary
    approval: ApprovalRequirementMetadata